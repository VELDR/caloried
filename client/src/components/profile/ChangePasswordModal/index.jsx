import PropTypes from 'prop-types';
import { useState } from 'react';
import { useDispatch } from 'react-redux';
import { useForm } from 'react-hook-form';
import { FormattedMessage, injectIntl } from 'react-intl';
import { Dialog, DialogContent } from '@mui/material';

import { Key, LockPerson, LockReset, Visibility, VisibilityOff } from '@mui/icons-material';
import SecondaryButton from '@components/ui/SecondaryButton';
import PrimaryButton from '@components/ui/PrimaryButton';
import { changePassword } from '@pages/Profile/actions';

import classes from './style.module.scss';

const ChangePasswordModal = ({ open, onClose, token, intl: { formatMessage } }) => {
  const dispatch = useDispatch();
  const {
    register,
    handleSubmit,
    formState: { errors },
    reset,
  } = useForm();

  const [showPassword, setShowPassword] = useState(false);

  const handleTogglePasswordVisibility = () => {
    setShowPassword(!showPassword);
  };

  const onSubmit = (data) => {
    dispatch(
      changePassword(data, token, () => {
        onClose();
        reset();
      })
    );
  };
  return (
    <Dialog open={open} onClose={onClose} maxWidth="xs" fullWidth>
      <DialogContent className={classes.dialog}>
        <div className={classes.dialog__title}>
          <LockReset /> <FormattedMessage id="app_change_password" />
        </div>
        <form onSubmit={handleSubmit(onSubmit)} className={classes.form}>
          <div className={classes.form__item}>
            <label htmlFor="currentPassword" className={classes.label}>
              <FormattedMessage id="app_current_password" />
              <LockPerson />
              <input
                id="currentPassword"
                type={showPassword ? 'text' : 'password'}
                className={`${classes.input} ${errors.currentPassword ? classes.inputError : ''}`}
                {...register('currentPassword', {
                  required: formatMessage({ id: 'app_current_password_is_required' }),
                })}
              />
              <div onClick={handleTogglePasswordVisibility} className={classes.eyeIcon}>
                {showPassword ? <VisibilityOff /> : <Visibility />}
              </div>
            </label>
            {errors.currentPassword && <span className={classes.error}>{errors.currentPassword.message}</span>}
          </div>
          <div className={classes.form__item}>
            <label htmlFor="newPassword" className={classes.label}>
              <FormattedMessage id="app_new_password" />
              <Key />
              <input
                id="newPassword"
                type={showPassword ? 'text' : 'password'}
                className={`${classes.input} ${errors.newPassword ? classes.inputError : ''}`}
                {...register('newPassword', {
                  required: formatMessage({ id: 'app_new_password_is_required' }),
                })}
              />
              <div onClick={handleTogglePasswordVisibility} className={classes.eyeIcon}>
                {showPassword ? <VisibilityOff /> : <Visibility />}
              </div>
            </label>
            {errors.newPassword && <span className={classes.error}>{errors.newPassword.message}</span>}
          </div>
          <div className={classes.buttons}>
            <SecondaryButton className={classes.button} onClick={onClose}>
              <FormattedMessage id="app_cancel" />
            </SecondaryButton>
            <PrimaryButton className={classes.button}>
              <div className={classes.text}>
                <FormattedMessage id="app_change" />
              </div>
            </PrimaryButton>
          </div>
        </form>
      </DialogContent>
    </Dialog>
  );
};

ChangePasswordModal.propTypes = {
  open: PropTypes.bool,
  onClose: PropTypes.func,
  intl: PropTypes.object,
  token: PropTypes.string,
};

export default injectIntl(ChangePasswordModal);
