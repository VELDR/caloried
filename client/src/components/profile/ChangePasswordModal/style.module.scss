@import '@styles/base';

.dialog {
  @include flex-column(1rem);
  &__title {
    @include flex-items-center(0.25rem);
    font-size: $font-size-large;
    font-weight: 500;
  }
  .form {
    @include flex-center-column(1rem);
    &__item {
      width: 75%;
      .label {
        position: relative;
        font-size: $font-size-regular;
        font-weight: 500;
        @include flex-column;
        svg {
          @include box(1rem);
          position: absolute;
          left: 8px;
          top: 67%;
          transform: translateY(-50%);
        }
        .input {
          @include input;
          padding-left: 1.75rem;
          padding-right: 1.75rem;
        }
        .inputError {
          border-color: $color-error;
          color: $color-error;
        }
        .eyeIcon {
          cursor: pointer;
          position: absolute;
          top: 65%;
          right: 14px;
          transform: translateY(-50%);
          @include box(1rem);
        }
      }
      .error {
        font-size: $font-size-smaller;
        color: $color-error;
      }
    }
    .buttons {
      @include flex-row(1rem);
      justify-content: flex-end;
      flex-wrap: wrap;
      align-self: flex-end;
      .button {
        @include flex-items-center(0.25rem);
      }
    }
  }
}
