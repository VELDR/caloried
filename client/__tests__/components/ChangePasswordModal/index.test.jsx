import ChangePasswordModal from '@components/profile/ChangePasswordModal';
import { fireEvent, render } from '@utils/testHelper';

let wrapper;

const mockProps = {
  open: true,
  onClose: jest.fn(),
  token: 'mockToken',
  intl: {
    formatMessage: jest.fn(),
  },
};

beforeEach(() => {
  wrapper = render(<ChangePasswordModal {...mockProps} />);
});

describe('ChangePasswordModal Component', () => {
  it('should render component correctly', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('change-password-modal')).toBeInTheDocument();
  });

  it('should toggle current password visibility', () => {
    const { getByLabelText, getByTestId } = wrapper;
    const currentPasswordInput = getByLabelText('Sandi sekarang');

    const toggleVisibilityCurrentPassword = getByTestId('toggle-visibility-current-password');

    expect(currentPasswordInput.getAttribute('type')).toBe('password');

    fireEvent.click(toggleVisibilityCurrentPassword);

    expect(currentPasswordInput.getAttribute('type')).toBe('text');
  });
  it('should toggle new password visibility', () => {
    const { getByLabelText, getByTestId } = wrapper;
    const newPasswordInput = getByLabelText('Sandi baru');

    const toggleVisibilityNewPassword = getByTestId('toggle-visibility-new-password');

    expect(newPasswordInput.getAttribute('type')).toBe('password');

    fireEvent.click(toggleVisibilityNewPassword);

    expect(newPasswordInput.getAttribute('type')).toBe('text');
  });

  it('should match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
