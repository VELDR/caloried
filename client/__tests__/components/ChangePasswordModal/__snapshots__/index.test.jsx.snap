// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangePasswordModal Component should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    />
    <div
      class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
      data-testid="change-password-modal"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-testid="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
        role="presentation"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        tabindex="-1"
      >
        <div
          aria-labelledby=":r3:"
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthXs MuiDialog-paperFullWidth css-twia2z-MuiPaper-root-MuiDialog-paper"
          role="dialog"
        >
          <div
            class="MuiDialogContent-root dialog css-ypiqx9-MuiDialogContent-root"
          >
            <div
              class="dialog__title"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="LockResetIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M13 3c-4.97 0-9 4.03-9 9H1l4 4 4-4H6c0-3.86 3.14-7 7-7s7 3.14 7 7-3.14 7-7 7c-1.9 0-3.62-.76-4.88-1.99L6.7 18.42C8.32 20.01 10.55 21 13 21c4.97 0 9-4.03 9-9s-4.03-9-9-9zm2 8v-1c0-1.1-.9-2-2-2s-2 .9-2 2v1c-.55 0-1 .45-1 1v3c0 .55.45 1 1 1h4c.55 0 1-.45 1-1v-3c0-.55-.45-1-1-1zm-1 0h-2v-1c0-.55.45-1 1-1s1 .45 1 1v1z"
                />
              </svg>
               
              Ubah Sandi
            </div>
            <form
              class="form"
            >
              <div
                class="form__item"
              >
                <label
                  class="label"
                  for="currentPassword"
                >
                  Sandi sekarang
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="LockPersonIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M18 11c.7 0 1.37.1 2 .29V10c0-1.1-.9-2-2-2h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h6.26c-.79-1.13-1.26-2.51-1.26-4 0-3.87 3.13-7 7-7zM8.9 6c0-1.71 1.39-3.1 3.1-3.1s3.1 1.39 3.1 3.1v2H8.9V6z"
                    />
                    <path
                      d="M18 13c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zm0 2c.83 0 1.5.67 1.5 1.5S18.83 18 18 18s-1.5-.67-1.5-1.5.67-1.5 1.5-1.5zm0 6c-1.03 0-1.94-.52-2.48-1.32.73-.42 1.57-.68 2.48-.68s1.75.26 2.48.68c-.54.8-1.45 1.32-2.48 1.32z"
                    />
                  </svg>
                  <input
                    class="input "
                    id="currentPassword"
                    name="currentPassword"
                    type="password"
                  />
                  <div
                    class="eyeIcon"
                    data-testid="toggle-visibility-current-password"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="VisibilityIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                      />
                    </svg>
                  </div>
                </label>
              </div>
              <div
                class="form__item"
              >
                <label
                  class="label"
                  for="newPassword"
                >
                  Sandi baru
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="KeyIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M21 10h-8.35C11.83 7.67 9.61 6 7 6c-3.31 0-6 2.69-6 6s2.69 6 6 6c2.61 0 4.83-1.67 5.65-4H13l2 2 2-2 2 2 4-4.04L21 10zM7 15c-1.65 0-3-1.35-3-3s1.35-3 3-3 3 1.35 3 3-1.35 3-3 3z"
                    />
                  </svg>
                  <input
                    class="input "
                    id="newPassword"
                    name="newPassword"
                    type="password"
                  />
                  <div
                    class="eyeIcon"
                    data-testid="toggle-visibility-new-password"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="VisibilityIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                      />
                    </svg>
                  </div>
                </label>
              </div>
              <div
                class="buttons"
              >
                <button
                  class="button button"
                  data-testid="secondary-button"
                  type="button"
                >
                  Batal
                </button>
                <button
                  class="button button"
                  data-testid="primary-button"
                  type="submit"
                >
                  <div
                    class="text"
                  >
                    Ubah
                  </div>
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
      <div
        data-testid="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
